core/src/SYCL/Kokkos_SYCL.cpp

```diff
@@ -76,11 +76,62 @@ int get_gpu(const InitArguments& args);
 }  // namespace Impl
 
 namespace Experimental {
-SYCL::SYCL() : m_space_instance(&Impl::SYCLInternal::singleton()) {
+SYCL::SYCL()
+    : m_space_instance(&Impl::SYCLInternal::singleton()), m_counter(nullptr) {
   Impl::SYCLInternal::singleton().verify_is_initialized(
       "SYCL instance constructor");
 }
 
+SYCL::SYCL(sycl::queue const stream)
+    : m_space_instance(new Impl::SYCLInternal), m_counter(new int(1)) {
+  Impl::SYCLInternal::singleton().verify_is_initialized(
+      "SYCL instance constructor");
+  m_space_instance->initialize(stream);
+}
+
+KOKKOS_FUNCTION SYCL::SYCL(SYCL&& other) noexcept {
+  m_space_instance       = other.m_space_instance;
+  other.m_space_instance = nullptr;
+  m_counter              = other.m_counter;
+  other.m_counter        = nullptr;
+}
+
+KOKKOS_FUNCTION SYCL::SYCL(const SYCL& other)
+    : m_space_instance(other.m_space_instance), m_counter(other.m_counter) {
+#ifndef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_SYCL
+  if (m_counter) Kokkos::atomic_add(m_counter, 1);
+#endif
+}
+
+KOKKOS_FUNCTION SYCL& SYCL::operator=(SYCL&& other) noexcept {
+  m_space_instance       = other.m_space_instance;
+  other.m_space_instance = nullptr;
+  m_counter              = other.m_counter;
+  other.m_counter        = nullptr;
+  return *this;
+}
+
+KOKKOS_FUNCTION SYCL& SYCL::operator=(const SYCL& other) {
+  m_space_instance = other.m_space_instance;
+  m_counter        = other.m_counter;
+#ifndef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_SYCL
+  if (m_counter) Kokkos::atomic_add(m_counter, 1);
+#endif
+  return *this;
+}
+
+KOKKOS_FUNCTION SYCL::~SYCL() noexcept {
```

Comment:
We should add comments to the class to the effect that:

If `m_counter` is set, then this instance is responsible for managing the objects pointed to by `m_counter` and `m_space_instance`.

This would be helpful for anyone in this code having to reason out if it is correct.